packages/block-editor/src/components/block-draggable/index.native.js

```diff
@@ -0,0 +1,315 @@
+				onDragOver={ updateDragging }
+				onDragEnd={ stopDragging }
+				wrapperAnimatedStyles={
+					styles[ 'draggable-wrapper__container' ]
+				}
+			>
+				{ children( { onScroll: scrollHandler } ) }
+			</Draggable>
+			<Animated.View
+				onLayout={ onChipLayout }
+				style={ chipStyles }
+				pointerEvents="none"
+			>
+				<DraggableChip />
+			</Animated.View>
+		</>
+	);
+};
+
+/**
+ * Block draggable component
+ *
+ * This component serves for animating the block when it is being dragged.
+ * Hence, it should be wrapped around the rendering of a block.
+ *
+ * @param {Object}      props          Component props.
+ * @param {JSX.Element} props.children Children to be rendered.
+ * @param {string[]}    props.clientId Client id of the block.
+ *
+ * @return {Function} Render function which includes the parameter `isDraggable` to determine if the block can be dragged.
+ */
+const BlockDraggable = ( { clientId, children } ) => {
+	const container = {
+		height: useSharedValue( 0 ),
+	};
+	const containerHeightBeforeDragging = useSharedValue( 0 );
+	const collapseAnimation = useSharedValue( 0 );
+
+	const onContainerLayout = ( { nativeEvent: { layout } } ) => {
+		container.height.value = layout.height;
+	};
+
+	const startBlockDragging = () => {
+		'worklet';
+		containerHeightBeforeDragging.value = container.height.value;
+		collapseAnimation.value = withTiming( 1 );
+	};
+
+	const stopBlockDragging = () => {
+		'worklet';
+		collapseAnimation.value = withTiming( 0 );
+	};
+
+	const { isDraggable, isBeingDragged } = useSelect(
+		( select ) => {
+			const {
+				getBlockRootClientId,
+				getTemplateLock,
+				isBlockBeingDragged,
+			} = select( blockEditorStore );
+			const rootClientId = getBlockRootClientId( clientId );
+			const templateLock = rootClientId
+				? getTemplateLock( rootClientId )
+				: null;
+
+			return {
+				isBeingDragged: isBlockBeingDragged( clientId ),
+				isDraggable: 'all' !== templateLock,
+			};
+		},
+		[ clientId ]
+	);
+
+	useEffect( () => {
+		if ( isBeingDragged ) {
+			runOnUI( startBlockDragging )();
+		} else {
+			runOnUI( stopBlockDragging )();
+		}
+	}, [ isBeingDragged ] );
+
+	const containerStyles = useAnimatedStyle( () => {
+		const height = interpolate(
+			collapseAnimation.value,
+			[ 0, 1 ],
+			[ containerHeightBeforeDragging.value, BLOCK_COLLAPSED_HEIGHT ]
+		);
+		return {
+			height:
+				containerHeightBeforeDragging.value === 0 ||
+				collapseAnimation.value === 0
+					? 'auto'
+					: height,
+		};
+	} );
+
+	const blockStyles = useAnimatedStyle( () => {
+		return {
+			opacity: 1 - collapseAnimation.value,
+		};
+	} );
+
+	const placeholderStyles = useAnimatedStyle( () => {
+		return {
+			display: collapseAnimation.value === 0 ? 'none' : 'flex',
+			opacity: collapseAnimation.value,
+		};
+	} );
+
+	if ( ! isDraggable ) {
+		return children( { isDraggable: false } );
+	}
+
+	return (
+		<Animated.View onLayout={ onContainerLayout } style={ containerStyles }>
```

Comment:
I'm wondering if we could try to get this data from the `blocksLayouts`. The size should be the same and it'd avoid having `onLayout` twice per block. What do you think?

Reply:
Good point, yep, I think we could save this `onLayout` listener by using the blocks layout data provided by the block list context 👍 . I'm going to do a quick check to verify that we receive the same values and if so, replace it with that data.

Reply:
I verified that we get the same value from the blocks layout data as `onLayout`, so I'll proceed to update this.

Reply:
I applied this suggestion in https://github.com/WordPress/gutenberg/pull/39617/commits/d16ce157ebc7b6e916c9c3026da48a525a71ca1b.

Reply:
Heads up that I had to update the collapse/expand animation due to these changes in https://github.com/WordPress/gutenberg/pull/39617/commits/bda7b248e6a583f497f75fd4578699c6db9a85f3. You can watch the new animation in the videos referenced in the PR's description.